smart-contracts/contracts/BondingCurveHook.sol

```diff
@@ -103,49 +82,55 @@ contract BondingCurveHook is ILockKeyPurchaseHookV7 {
   ) external view
     returns (uint minKeyPrice)
   {
-    // Not needed for this use-case.
+    // no-op.
   }
 
+  // https://ethereum.stackexchange.com/questions/15350/how-to-convert-an-bytes-to-address-in-solidity
+  function bytesToAddress(bytes memory b) private pure returns (address addr) {
+    assembly {
+      addr := mload(add(b,20))
+    }
+}
+
   /**
    * @notice If the lock owner has registered an implementer then this hook
    * is called with every key sold.
-   * @param from the msg.sender making the purchase
-   * @param recipient the account which will be granted a key
-   * @param referrer the account which referred this key sale
+   *  from the msg.sender making the purchase
+   *  recipient the account which will be granted a key
+   *  referrer the account which referred this key sale
    * @param data arbitrary data populated by the front-end which initiated the sale
-   * @param minKeyPrice the price including any discount granted from calling this
+   *  minKeyPrice the price including any discount granted from calling this
    * hook's `keyPurchasePrice` function
-   * @param pricePaid the value/pricePaid included with the purchase transaction
+   *  pricePaid the value/pricePaid included with the purchase transaction
    * @dev the lock's address is the `msg.sender` when this function is called
    */
   function onKeyPurchase(
-    address from,
-    address recipient,
-    address referrer,
+    address /*from**/,
+    address /*recipient**/,
+    address /*referrer**/,
     bytes calldata data,
-    uint minKeyPrice,
-    uint pricePaid
+    uint /*minKeyPrice**/,
+    uint /*pricePaid**/
   ) external
   {
-    // Ensure caller is the locked.fyi lock
-    require(msg.sender == LOCK_ADDRESS);
+    // Ensure caller is the locked.fyi lock:
+    // @audit re-enable check before deployment !!!
+    // require(msg.sender == LOCK_ADDRESS);
+    require(msg.sender == _testLockAddress);
     IPublicLockV7 lock = IPublicLockV7(msg.sender);
-    // @audit sort out units
-    int128 supply = tokenSupply.fromUInt();
-    int128 logTokenPrice = supply.log_2();
-    int128 modifiedTokenPrice = logTokenPrice.div(CURVE_MODIFER).div(DENOMINATOR);
-    uint256 keyPrice = modifiedTokenPrice.toUInt();
-    // update current supply using internal counter
     tokenSupply++;
 
-    // Read token address from lock and pass as 2nd arg:
+    // calculate the price for the new supply:
+    int128 supply = tokenSupply.fromUInt();
+    uint keyPrice = supply.log_2().div(CURVE_MODIFER).div(DENOMINATOR).mulu(1 * 10 ** 18);
+
+    // get current token address from lock:
     address tokenAddress = lock.tokenAddress();
     lock.updateKeyPricing(keyPrice, tokenAddress);
```

Comment:
I think we said we should only do that so often to avoid concurrency issues?

Reply:
We have a couple of options t consider.
 We discussed a periodic price increase rather than continuous. I have a few thoughts on this:
- price increase only every so often could save 5000 gas per purchase (for most of the txs)
- mechanism used to track steps in function would have to use < 5000 gas for savings, while adding complexity.
- maybe accounting for price slippage could solve the concurrency issues.
Rather than adding to the smart contract, the concurrency issue could maybe be dealt with at the UI level. Calculate the price of the current key (based on the current supply), as well as the next key, and approve enough extra to buy a slightly more expensive key if necessary.
- Established pattern (uniswap)
- the extra amount needed to purchase the next key instead of the current one rapidly becomes negligible.
  eg: by the 10th key sold, the price is increasing by only 5% per key. By the 50th key, only 1%, and is under 0.5 % by the 75th key.

Reply:
I opened a new issue to track this, #54